README.md
# Rank-biserial correlation

RBCDE is a Python implementation of the rank-biserial correlation coefficient [(Cureton, 1956)](https://link.springer.com/article/10.1007/BF02289138), which can be used as an effect size equivalent of the Wilcoxon test [(Kerby, 2014)](https://journals.sagepub.com/doi/full/10.2466/11.IT.3.1), which in turn was deemed to perform well on single cell data problems [(Soneson, 2018)](https://www.nature.com/articles/nmeth.4612). Using effect size analyses is recommended for problems with large population sizes [(Sullivan, 2012)](https://www.jgme.org/doi/full/10.4300/JGME-D-12-00156.1). The package comes with both a [scanpy](https://scanpy.readthedocs.io/en/latest/)-compatible version and a standalone function that ingests a data matrix and an assignment vector.

## Citation

Stay tuned!

## Installation

RBCDE depends on numpy, scipy and pandas. The package is available on pip, and can be easily installed as follows:

	pip3 install rbcde

## Usage and Documentation

RBCDE can slot into a scanpy workflow and accept an object with per cell normalised data stored as a layer or `.raw`, and the desired clustering/grouping vector as an `.obs` column. If mirroring the [scanpy PBMC tutorial](https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html), the output of `sc.pp.log1p()` produces a generally useful data representation (the earlier step `sc.pp.normalize_total()` will yield identical RBCDE results).

	import rbcde
	rbcde.RBC(adata)
	degs, plot_dict = rbcde.filter_markers(adata)

`rbcde.RBC()`'s `clus_key` argument controls which `.obs` column is used for the grouping, and a combination of `layer` and `use_raw` can instruct the function to retrieve expression data from `.X`, `.layers` or `.raw`. `rbcde.filter_markers()` takes the computed coefficient values and thresholds them into a list of per-cluster markers. The thresholding can be controlled via the `thresh` argument, with a range of literature [critical values](https://en.wikipedia.org/wiki/Effect_size#Pearson_r_or_correlation_coefficient) available. A helper dictionary, compatible with the formatting scanpy plotting functions accept in the `var_names` argument.

Analogous functions exist for scanpy-independent data analysis, and can ingest any data matrix with variables as rows and observations as columns along with a vector of cluster/group assignments for the observations and a second vector of variable names. The filtering function does not produce a helper dictionary, only yielding the marker data frame.

	results = rbcde.matrix.RBC(data, clusters, vars)
	degs = rbcde.matrix.filter_markers(results)

An HTML render of the RBCDE function docstrings, detailing all the parameters, can be accessed at [ReadTheDocs](https://rbcde.readthedocs.io/en/latest/).

## Example Notebook

[rbc_demo.ipynb](https://nbviewer.jupyter.org/github/Teichlab/rbcde/blob/master/examples/rbc_demo.ipynb) computes the rank-biserial correlation coefficient for demonstration 10X PBMC data, yielding a similar standard of markers to established approaches while reporting only ~13% of the gene total. This more compact summary does not require any heuristic filtering to obtain. The full marker export yielded by the analysis can be found at `examples/markers.csv`

CHANGELOG.md
## 1.0.0
- initial release

rbcde/matrix.py
import pandas as pd
import numpy as np
from scipy import stats, sparse

def RBC(data, clusters, vars):
	'''
	Compute the rank-biserial correlation coefficient for each gene in each cluster. The 
	results can be subsequently turned into a marker list via the helper function 
	``rbcde.matrix.filter_markers()``. Returns a data frame with the coefficient value for 
	each gene in each cluster.
	
	The rank-biserial correlation coefficient 
	`(Cureton, 1956) <https://link.springer.com/article/10.1007/BF02289138>`_ 
	can be used as an effect size equivalent of the Wilcoxon test 
	`(Kerby, 2014) <https://journals.sagepub.com/doi/full/10.2466/11.IT.3.1>`_. Using effect 
	size analyses is recommended for problems with large population sizes 
	`(Sullivan, 2012) <https://www.jgme.org/doi/full/10.4300/JGME-D-12-00156.1>`_.
	
	Input
	-----
	data : ``np.array`` or ``scipy.sparse``
		Per cell normalised, if using single cell count data. Variables as rows, 
		observations as columns.
	clusters : ``np.array`` or ``list``
		A vector of cluster/group assignments for each observation.
	vars: ``np.array`` or ``list``
		A vector of variable names, for output generation purposes.
	'''
	
	#set up sparse data as CSC format for quickness of access
	#and a little helper flag to see if we need to densify or not
	issparse = False
	if sparse.issparse(data):
		data = data.tocsc()
		issparse = True
	#prepare some cluster'y stuff
	clusters_categories = np.unique(clusters)
	#store the boolean mask of cluster membership, nU1 (number of cells in each cluster)
	#and nU1*nU2 (multiply that by the number of cells not in that cluster)
	masks = {}
	nU1 = {}
	nU1nU2 = {}
	for clus in clusters_categories:
		masks[clus] = np.array([i==clus for i in clusters])
		nU1[clus] = np.sum(masks[clus])
		nU1nU2[clus] = nU1[clus]*np.sum(~masks[clus])
	#make results array
	results = np.zeros((len(vars),len(clusters_categories)))
	#okay, with all this out of the way, can finally get computing
	for i in np.arange(len(vars)):
		#pre-compute the gene's ranks once
		if issparse:
			expr = data[:,i].todense()
		else:
			expr = data[:,i]
		ranks = stats.rankdata(expr)
		for j, clus in enumerate(clusters_categories):
			#compute the actual RBC coefficient and store it
			#making active use of the fact U1+U2 = nU1*nU2
			U1 = np.sum(ranks[masks[clus]]) - nU1[clus]*(nU1[clus]+1)/2
			U2 = nU1nU2[clus] - U1
			results[i,j] = (U1-U2)/nU1nU2[clus]
	return(pd.DataFrame(results, index=vars, columns=clusters_categories))

def filter_markers(results, thresh=0.5):
	'''
	Filter the rank-biserial correlation coefficients computed with ``rbcde.matrix.RBC()`` to a 
	list of markers for each cluster. Returns a data frame of the computed markers.
	
	Input
	-----
	results : ``pd.DataFrame``
		The output of ``rbcde.matrix.RBC()``.
	thresh : ``float``, optional (default: 0.5)
		The threshold value used to call markers. Literature 
		`critical values <https://en.wikipedia.org/wiki/Effect_size#Pearson_r_or_correlation_coefficient>`_ 
		can be used.
	'''
	
	degs = pd.DataFrame(columns=['cluster','RBC'])
	for clus in results.columns:
		sub = pd.DataFrame(clus, index=results.index, columns=['cluster','RBC'])
		sub['RBC'] = results[clus]
		degs = degs.append(pd.DataFrame(sub[sub['RBC']>thresh]).sort_values(by='RBC', ascending=False))
	return(degs)

rbcde/__init__.py
import rbcde.matrix
import pandas as pd
try:
	from scanpy import logging as logg
except ImportError:
	pass

def RBC(adata, clus_key='leiden', layer=None, use_raw=False):
	'''
	Compute the rank-biserial correlation coefficient for each gene in each cluster. The 
	results can be subsequently turned into a marker list via the helper function 
	``rbcde.filter_markers()``. The primary output is stored as part of either `.var` or 
	`.raw.var`, depending on whether `.raw` data is used.
	
	The rank-biserial correlation coefficient 
	`(Cureton, 1956) <https://link.springer.com/article/10.1007/BF02289138>`_ 
	can be used as an effect size equivalent of the Wilcoxon test 
	`(Kerby, 2014) <https://journals.sagepub.com/doi/full/10.2466/11.IT.3.1>`_, which in 
	turn was deemed to perform well on single cell data problems 
	`(Soneson, 2018) <https://www.nature.com/articles/nmeth.4612>`_. Using effect size 
	analyses is recommended for problems with large population sizes 
	`(Sullivan, 2012) <https://www.jgme.org/doi/full/10.4300/JGME-D-12-00156.1>`_.
	
	Input
	-----
	adata : ``AnnData``
		Needs per cell normalised data stored somewhere in the object (as either sparse or 
		dense), and the desired clustering/grouping vector included in `.obs`.
	clus_key : ``str``, optional (default: "leiden")
		The name of the `.obs` column containing the clustering/grouping.
	layer : ``str`` or ``None``, optional (default: ``None``)
		If specified, take the expression data from the matching ``.layers`` field. Overrides 
		``use_raw`` if provided.
	use_raw : ``bool``, optional (default: ``False``)
		If no ``layer`` was specified and this is set to ``True``, take the data from the 
		``.raw`` field of the object. Store results in ``.raw.var`` to match dimensionality.
	'''
	
	start = logg.info('computing rank-biserial correlation')
	#extract appropriate data form, along with gene names
	#layer trumps use_raw
	raw_prefix = ''
	if layer is not None:
		data = adata.layers[layer]
		genes = adata.var_names
	else:
		#try to get out .raw if possible (and specified)
		if adata.raw is not None and use_raw:
			data = adata.raw.X.copy()
			genes = adata.raw.var_names
			raw_prefix = '.raw'
		else:
			data = adata.X.copy()
			genes = adata.var_names
	#extract cluster list
	clusters = adata.obs[clus_key].copy()
	#think that's everything we need. call matrix version of function
	rbc_out = rbcde.matrix.RBC(data, clusters, genes)
	#append the fact this is RBC output to each column name and stash it in var
	for col in rbc_out.columns:
		if use_raw:
			adata.raw.var['RBC_'+col] = rbc_out[col]
		else:
			adata.var['RBC_'+col] = rbc_out[col]
	logg.info('	finished', time=start,
		deep=(
			'added\n'
			"	'RBC_' columns for each of the clusters to "+raw_prefix+".var"
		),)

def filter_markers(adata, thresh=0.5, use_raw=False):
	'''
	Filter the rank-biserial correlation coefficients computed with ``rbcde.RBC()`` to a 
	list of markers for each cluster, provided as a data frame and a Scanpy plotting compatible 
	``var_names`` cluster marker dictionaty. Returns those two objects, in this order.
	
	Input
	-----
	adata : ``AnnData``
		Needs to have been processed with ``rbcde.RBC()``.
	thresh : ``float``, optional (default: 0.5)
		The threshold value used to call markers. Literature 
		`critical values <https://en.wikipedia.org/wiki/Effect_size#Pearson_r_or_correlation_coefficient>`_ 
		can be used.
	use_raw : ``bool``, optional (default: ``False``)
		Set this to ``True`` if the raw data was used for the computation so that the 
		results can be retrieved from the correct field of the object.
	'''
	
	
	#extract the RBC results embedded in .var and remove the prefix
	if use_raw:
		results = adata.raw.var.loc[:,[i.startswith('RBC_') for i in adata.raw.var.columns]]
	else:
		results = adata.var.loc[:,[i.startswith('RBC_') for i in adata.var.columns]]
	results.columns = [i.replace('RBC_','',1) for i in results.columns]
	#call the matrix version to get a marker data frame
	degs = rbcde.matrix.filter_markers(results, thresh)
	#parse up a plotting cluster marker dictionary
	plot_dict = {}
	for clus in results.columns:
		plot_dict[clus] = degs.loc[degs['cluster']==clus, :].index
		logg.hint(str(len(plot_dict[clus]))+' markers found for cluster '+clus)
	#return both the data frame and the plot-ready form
	return degs, plot_dict